<a href="https://colab.research.google.com/github/Xerxes80/CS-6301.M02/blob/main/Text_Classification_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

About Dataset
IMDB dataset having 50K movie reviews for natural language processing or Text analytics.
This is a dataset for binary sentiment classification containing substantially more data than previous benchmark datasets. We provide a set of 25,000 highly polar movie reviews for training and 25,000 for testing. So, predict the number of positive and negative reviews using either classification or deep learning algorithms.

In [ ]:
import pandas as pd
# For Naive Bayes
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score
# For Logistic Regression
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
# For Neural Networks 
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score


# Load the dataset

train_df = pd.read_csv('/content/imdb_train.csv', header=0, lineterminator='\n')
train_df = train_df.applymap(lambda x: x.strip() if isinstance(x, str) else x)

nRow, nCol = train_df.shape
print(f'There are {nRow} rows and {nCol} columns in the training set')

# Removing empty rows from csv 
train_df.dropna(axis=0, how='all',inplace=True)
nRow, nCol = train_df.shape
print(f'There are {nRow} rows and {nCol} columns in the training set after removing empty rows')

train_df.head()

There are 50000 rows and 2 columns in the training set
There are 50000 rows and 2 columns in the training set after removing empty rows


review sentiment
0  One of the other reviewers has mentioned that ...  positive
1  A wonderful little production. <br /><br />The...  positive
2  I thought this was a wonderful way to spend ti...  positive
3  Basically there's a family where a little boy ...  negative
4  Petter Mattei's "Love in the Time of Money" is...  positive

In [ ]:
test_df = pd.read_csv('/content/imdb_test.csv', header=0, lineterminator='\n')
test_df = test_df.applymap(lambda x: x.strip() if isinstance(x, str) else x)

nRow, nCol = test_df.shape
print(f'There are {nRow} rows and {nCol} columns in the testing set')

# Removing empty rows from csv 
test_df.dropna(axis=0, how='all',inplace=True)
nRow, nCol = test_df.shape
print(f'There are {nRow} rows and {nCol} columns in the testing set after removing empty rows')

test_df.head()

There are 50000 rows and 2 columns in the testing set
There are 50000 rows and 2 columns in the testing set after removing empty rows


review sentiment
0  One of the other reviewers has mentioned that ...  positive
1  A wonderful little production. <br /><br />The...  positive
2  I thought this was a wonderful way to spend ti...  positive
3  Basically there's a family where a little boy ...  negative
4  Petter Mattei's "Love in the Time of Money" is...  positive

In [ ]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review     50000 non-null  object
 1   sentiment  50000 non-null  object
dtypes: object(2)
memory usage: 781.4+ KB


In [ ]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review     50000 non-null  object
 1   sentiment  50000 non-null  object
dtypes: object(2)
memory usage: 781.4+ KB


In [ ]:
train_df.isnull().sum()

review       0
sentiment    0
dtype: int64

In [ ]:
test_df.isnull().sum()

review       0
sentiment    0
dtype: int64

In [ ]:
# Extract the features and labels
X_train = train_df["review"]
y_train = train_df["sentiment"]
X_test = test_df["review"]
y_test = test_df["sentiment"]

# Vectorize the text data
vectorizer = CountVectorizer(stop_words="english")
X_train_vectorized = vectorizer.fit_transform(X_train)
X_test_vectorized = vectorizer.transform(X_test)

# Train a Naive Bayes classifier
clf = MultinomialNB()
clf.fit(X_train_vectorized, y_train)

# Make predictions on the test set
y_pred = clf.predict(X_test_vectorized)

# Evaluate the accuracy of the classifier
accuracy = accuracy_score(y_test, y_pred)
print("Naive Bayes Accuracy:", accuracy)

Naive Bayes Accuracy: 0.8983


In [ ]:

# Vectorize the text data
vectorizer = TfidfVectorizer(stop_words="english", max_features=5000)
X_train_vectorized = vectorizer.fit_transform(X_train)
X_test_vectorized = vectorizer.transform(X_test)

# Train a Logistic Regression classifier
clf = LogisticRegression(max_iter=1000)
clf.fit(X_train_vectorized, y_train)

# Make predictions on the test set
y_pred = clf.predict(X_test_vectorized)

# Evaluate the accuracy of the classifier
accuracy = accuracy_score(y_test, y_pred)
print("Logistic Regression Accuracy:", accuracy)

Logistic Regression Accuracy: 0.91062


In [ ]:
# Vectorize the text data
vectorizer = TfidfVectorizer(stop_words="english", max_features=5000)
X_train_vectorized = vectorizer.fit_transform(X_train)
X_test_vectorized = vectorizer.transform(X_test)

# Train a Neural Network classifier
clf = MLPClassifier(hidden_layer_sizes=(100,), max_iter=1000)
clf.fit(X_train_vectorized, y_train)

# Make predictions on the test set
y_pred = clf.predict(X_test_vectorized)

# Evaluate the accuracy of the classifier
accuracy = accuracy_score(y_test, y_pred)
print("Neural Network Accuracy:", accuracy)

Neural Network Accuracy: 0.9999


The Naive Bayes approach achieved an accuracy of 0.8983. This method is relatively simple and works well with high-dimensional sparse data. However, it makes a strong assumption of independence between the features, which may not hold true in real-world scenarios.

The Logistic Regression approach achieved an accuracy of 0.91062. This method is a popular choice for binary classification problems and works well with large datasets. It is also relatively easy to interpret the coefficients of the model. However, it may not perform well when the data has non-linear relationships.

The Neural Network approach achieved an accuracy of 0.9999. This method is a powerful tool for learning complex non-linear relationships in data. It can also handle high-dimensional data and is known to perform well on text data. However, it requires more computational resources and may be more difficult to interpret.

Overall, the three approaches performed well in predicting the sentiment of the movie review dataset. The Naive Bayes approach is a good choice for its simplicity and speed, while the Logistic Regression approach is a solid option for large datasets with binary classification problems. The Neural Network approach is a powerful tool for learning complex relationships in data, but may require more computational resources and expertise in model tuning.
